### PR TITLE
Handle duplicated outputs in eval mode and under torch.no_grad context

### PR DESCRIPTION
**Pull request by baijumeswani**
In ```eval``` mode and under ```torch.no_grad()``` context, ```ORTModule``` uses the ```inference_agent``` which uses ```IOBinding``` to execute the graph. ```IOBinding``` only registers a single output for duplicate output names. As a result, when the model has duplicated outputs, the returned values from the ORT backend will only correspond to the ones registered through ```IOBinding```.

This pull request introduces a mapping for output names to the index of the ORT outputs. And so adds support for duplicated outputs in the user model.

### CODE REVIEW

**Review by thiagocrepaldi — COMMENTED** [review]
Is it possible to stop using IOBinding altogether? Having maps being constructed for every forward call can be an additional burden that will pop up on future benchmark

Technically speaking, maybe we can stop using IO binding and use DLPack. @iK1D may comment on that

**Comment by baijumeswani** [comment]
The mapping is not created on every forward call. It is cached and only created when the model is exported.
I can try to see if there is a way to do away with ```IOBinding```.

**Comment by thiagocrepaldi** [comment]
> The mapping is not created on every forward call. It is cached and only created when the model is exported.
> I can try to see if there is a way to do away with `IOBinding`.

Thanks, fixing or replacing IOBinding would be a better fix as the code is under our control, differently from other issues related to PyTorch or ONNX that we handle on frontend

**Comment by baijumeswani** [comment]
Opened another pull request to address this issue through graph transformations: https://github.com/microsoft/onnxruntime/pull/8654. Closing this.
